File: test/_helpers/repo_fixture.dart
import 'dart:io';
import 'package:path/path.dart' as p;

class RepoFixture {
  RepoFixture._(this.path, this.headSha);
  final String path;
  String headSha;

  static Future<RepoFixture> empty() async {
    final dir = Directory.systemTemp.createTempSync('gitopen-test-');
    await _git(dir.path, ['init', '-q', '-b', 'master']);
    await _git(dir.path, ['config', 'user.email', 'test@example.com']);
    await _git(dir.path, ['config', 'user.name', 'Test']);
    await _git(dir.path, ['config', 'commit.gpgsign', 'false']);
    return RepoFixture._(dir.path, '');
  }

  static Future<RepoFixture> withLinearHistory(int commits) async {
    if (commits < 1) throw ArgumentError.value(commits, 'commits');
    final f = await empty();
    for (var i = 0; i < commits; i++) {
      final file = File(p.join(f.path, 'file_$i.txt'));
      await file.writeAsString('content $i\n');
      await _git(f.path, ['add', 'file_$i.txt']);
      await _git(f.path, ['commit', '-q', '-m', 'commit $i']);
    }
    final headOut = await _git(f.path, ['rev-parse', 'HEAD']);
    f.headSha = headOut.trim();
    return f;
  }

  static Future<RepoFixture> withFileRemote() async {
    final origin = await withLinearHistory(3);
    final local = await empty();
    await _git(local.path, ['remote', 'add', 'origin', origin.path]);
    return local;
  }

  static Future<RepoFixture> withBranches() async {
    final f = await withLinearHistory(3);
    await _git(f.path, ['checkout', '-q', '-b', 'feature']);
    final file = File(p.join(f.path, 'feature.txt'));
    await file.writeAsString('feature\n');
    await _git(f.path, ['add', 'feature.txt']);
    await _git(f.path, ['commit', '-q', '-m', 'on feature']);
    await _git(f.path, ['checkout', '-q', 'master']);
    return f;
  }

  /// Repo whose HEAD is a real (`--no-ff`) merge commit with two parents.
  /// `master` adds master.txt, `feature` adds feature.txt; the merge
  /// introduces feature.txt relative to its FIRST parent (master).
  static Future<RepoFixture> withMergeCommit() async {
    final f = await empty();
    await File(p.join(f.path, 'file_0.txt')).writeAsString('content 0\n');
    await _git(f.path, ['add', 'file_0.txt']);
    await _git(f.path, ['commit', '-q', '-m', 'commit 0']);

    await _git(f.path, ['checkout', '-q', '-b', 'feature']);
    await File(p.join(f.path, 'feature.txt')).writeAsString('feature\n');
    await _git(f.path, ['add', 'feature.txt']);
    await _git(f.path, ['commit', '-q', '-m', 'on feature']);

    await _git(f.path, ['checkout', '-q', 'master']);
    await File(p.join(f.path, 'master.txt')).writeAsString('master\n');
    await _git(f.path, ['add', 'master.txt']);
    await _git(f.path, ['commit', '-q', '-m', 'on master']);

    await _git(f.path, [
      'merge',
      '-q',
      '--no-ff',
      '-m',
      'merge feature',
      'feature',
    ]);
    f.headSha = (await _git(f.path, ['rev-parse', 'HEAD'])).trim();
    return f;
  }

  Future<void> dispose() async {
    try {
      await Directory(path).delete(recursive: true);
    } on Object {
      // Best-effort cleanup; ignore failures (e.g. locked files on Windows).
    }
  }

  static Future<String> _git(String cwd, List<String> args) async {
    final r = await Process.run('git', args, workingDirectory: cwd);
    if (r.exitCode != 0) {
      throw StateError('git ${args.join(' ')} failed: ${r.stderr}');
    }
    return r.stdout.toString();
  }
}

File: tests/GitOpen.Infrastructure.Tests/Helpers/RepoFixture.cs
using LibGit2Sharp;

namespace GitOpen.Infrastructure.Tests.Helpers;

public sealed class RepoFixture : IDisposable
{
    public string Path { get; }
    public string HeadSha { get; private set; } = "";

    private RepoFixture(string path) { Path = path; }

    public static RepoFixture Empty()
    {
        var path = CreateTempPath();
        Repository.Init(path);
        return new RepoFixture(path);
    }

    public static RepoFixture WithLinearHistory(int commits)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(commits, 1);
        var fixture = Empty();
        using var repo = new Repository(fixture.Path);
        var sig = new Signature("Test", "test@example.com", DateTimeOffset.UtcNow);
        for (var i = 0; i < commits; i++)
        {
            var file = System.IO.Path.Combine(fixture.Path, $"file_{i}.txt");
            File.WriteAllText(file, $"content {i}\n");
            Commands.Stage(repo, $"file_{i}.txt");
            var c = repo.Commit($"commit {i}", sig, sig);
            fixture.HeadSha = c.Sha;
        }
        return fixture;
    }

    public static RepoFixture WithBranches()
    {
        var fixture = WithLinearHistory(3);
        using var repo = new Repository(fixture.Path);
        var sig = new Signature("Test", "test@example.com", DateTimeOffset.UtcNow);
        var feature = repo.CreateBranch("feature");
        Commands.Checkout(repo, feature);
        var file = System.IO.Path.Combine(fixture.Path, "feature.txt");
        File.WriteAllText(file, "feature\n");
        Commands.Stage(repo, "feature.txt");
        repo.Commit("on feature", sig, sig);
        Commands.Checkout(repo, repo.Branches["master"] ?? repo.Branches["main"]!);
        return fixture;
    }

    public void Dispose()
    {
        try { ForceDelete(Path); } catch { /* best-effort cleanup */ }
    }

    private static string CreateTempPath()
    {
        var p = System.IO.Path.Combine(System.IO.Path.GetTempPath(), "gitopen-test-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(p);
        return p;
    }

    private static void ForceDelete(string path)
    {
        if (!Directory.Exists(path)) return;
        foreach (var f in Directory.GetFiles(path, "*", SearchOption.AllDirectories))
            File.SetAttributes(f, FileAttributes.Normal);
        Directory.Delete(path, recursive: true);
    }
}
